### DISCUSSION TITLE
How to map a C# enum to a PostgreSQL enum type?

### DISCUSSION THREAD

**Discussion by abirmkj** [MoonStorm/FastCrud#164] [Q&A]
I have a C# class `Sample`, one of whose properties is an enum, `TargetType`. I have a corresponding table `samples` defined in a PostgreSQL database, along with a matching enum type, `targettypes`.

With Dapper.FastCRUD, I can retrieve records from the table successfully. However, I get an error during insertion:

`Npgsql.PostgresException (0x80004005): 42804: column "target_type" is of type targettype but expression is of type integer`

**How do I tell Dapper.FastCRUD to map the C# enum TargetType to PostgreSQL targettype for inserts?**

In my C# code, I've defined:

```
// Map the C# enum 'TargetType' to PostgreSQL enum 'targettype'
// Allows Npgsql to insert and retrieve the fields with correct type-casting / boxing
Npgsql.NpgsqlConnection.MapEnumGlobally<TargetType>("public.targettype");  

Dapper.FastCrud.OrmConfiguration.DefaultDialect = Dapper.FastCrud.SqlDialect.PostgreSql;
```

When using only Npgsql (and no ORM), I can successfully insert as follows:

```
using (var writer = conn.BeginBinaryImport(sql))
{
    foreach (Sample sample in entities)
    {
        writer.StartRow();
        ...
        writer.Write(sample.TargetType); // <--- mapped to column 'target_type'
        ...
    }
}
```

**Comment by MoonStorm** [comment]
No DB-CLR type conversions occur at the level of this library. We're passing and retrieving data through Dapper.

**Comment by abirmkj** [comment]
Thanks for the clarification @MoonStorm . 
Could you point me to some resources that would help resolve this issue?

**Comment by MoonStorm** [comment]
While in this particular case I'd recommend sticking to the basic primitive types (ie use an integer instead of an enum), I've started  a section in the wiki about value mappings [here](https://github.com/MoonStorm/FastCrud/wiki/Property-value-mappings). We can expand it later with other examples.
